sdk/src/utils/graphs/adjacency-pool-graph.ts

```diff
@@ -0,0 +1,195 @@
+import { AddressUtil } from "@orca-so/common-sdk";
+import { Address } from "@project-serum/anchor";
+import { PublicKey } from "@solana/web3.js";
+import {
+  Hop,
+  PoolGraph,
+  PoolGraphUtils,
+  PoolTokenPair,
+  Route,
+  RouteFindOptions,
+  RouteSearchEntires
+} from "../public/pool-graph";
+
+export class AdjacencyPoolGraph implements PoolGraph {
+  readonly graph: {
+    [k: string]: PoolGraphEdge[];
+  };
+
+  constructor(pools: PoolTokenPair[]) {
+    this.graph = buildPoolGraph(pools);
+  }
+
+  getRoute(startMint: Address, endMint: Address, options?: RouteFindOptions): Route[] {
+    const [startMintKey, endMintKey] = [
+      AddressUtil.toString(startMint),
+      AddressUtil.toString(endMint),
+    ];
+
+    const walkMap = findWalks(
+      [[startMintKey, endMintKey]],
+      this.graph,
+      options?.intermediateTokens.map((token) => AddressUtil.toString(token))
+    );
+
+    const internalRouteId = getInternalRouteId(startMintKey, endMintKey);
+    const searchRouteId = PoolGraphUtils.getSearchRouteId(startMint, endMint);
+    const routeForSearchPair = walkMap[internalRouteId];
+    if (routeForSearchPair === undefined) {
+      return [];
+    }
+
+    return routeForSearchPair.map((route) => {
+      return {
+        startTokenMint: startMintKey,
+        endTokenMint: endMintKey,
+        hops: getHopsFromRoute(internalRouteId, searchRouteId, route),
+      };
+    });
+  }
+
+  getAllRoutes(searchTokenPairs: [Address, Address][], options?: RouteFindOptions): RouteSearchEntires {
+    const searchTokenPairsAddrs = searchTokenPairs.map(([startMint, endMint]) => {
+      return [AddressUtil.toString(startMint), AddressUtil.toString(endMint)] as const;
+    });
+    const walkMap = findWalks(
+      searchTokenPairsAddrs,
+      this.graph,
+      options?.intermediateTokens.map((token) => AddressUtil.toString(token))
+    );
+
+    const results = searchTokenPairsAddrs.map(([startMint, endMint]) => {
+      const searchRouteId = PoolGraphUtils.getSearchRouteId(startMint, endMint);
+      const internalRouteId = getInternalRouteId(startMint, endMint);
+      const routesForSearchPair = walkMap[internalRouteId];
+      const paths = routesForSearchPair ? routesForSearchPair.map<Route>((route) => {
+        return {
+          startTokenMint: startMint,
+          endTokenMint: endMint,
+          hops: getHopsFromRoute(internalRouteId, searchRouteId, route),
+        };
+      }) : [];
+
+
+      return [searchRouteId, paths] as const;
+    });
+
+    return results;
+  }
+}
+
+
+
+function getHopsFromRoute(internalRouteId: string, searchRouteId: string, route: string[]): Hop[] {
+  const [intStartA] = PoolGraphUtils.deconstructRouteId(internalRouteId);
+  const [searchStartA] = PoolGraphUtils.deconstructRouteId(searchRouteId);
+  const shouldReverseRoute = searchStartA !== intStartA;
+  const finalRoutes = shouldReverseRoute ? route.reverse() : route;
```

Comment:
is ``route.slice().reverse()``  safer ?

In my understanding, .reverse() is in-place method, so I am concerned walkMap[internalRouteId] is also reversed.
I am concerned about whether a problem will arise if the reversal order is made first and the nonreversal order comes later.
edge case input: [[A, B], [B, A]] and [[B, A], [A, B]]

Reply:
good catch!